### PR TITLE
Fix gazebo lidar sensor output to ROS topic

### LINKED ISSUE #325: GPU Lidar sensor bridge is not passing all of the sensor data from the GZ topic to the defined ROS2 topic.
## Environment
* OS Version: Ubuntu 22.04
ROS2: Humble, binary install
Gazebo: Gazebo Garden, binary install
ros_ign branch: ros2 (master branch, re-cloned today)
## Description
* Expected behavior: When creating the bridge between the GPU_Lidar sensor and Gazebo Garden, I would expect the horizontal and vertical samples from the sesnor to be passed to the defined Ros2 topic.
* Actual behavior: When I try to pass the sensor measurements to Ros2 the only lidar measurements being received are from the first vertical sample and that array of horizontal samples. For example, if I define the GPU Lidar to have 10 horizontal and 10 vertical samples, it generates 100 total Lidar points. However, I am only getting 10 of them passed to ROS. 

## Steps to reproduce
1. Make a world with the GPU lidar sensor
2. In the launch file for the world, use the bridge to send the gazebo topic data to ROS.
3. With some parameter changes it will be noticed that increasing the number of vertical samples increases the number of ranges in the gz topic, but not the number of ranges in the ROS2 topic. 

## Output
Here I will prove images of the SDF file for my sensor, the bridge in my launch file, and the GZ topic console output vs the ROS2 topic console output.
image
image
image
image

### PR DESCRIPTION
**Pull request by ConnorDTaylor**
# 🦟 Bug fix

Fixes #325

## Summary
I noticed that the GPU LIDAR sensor would only pass one horizontal line of ranges and intensities from Gazebo to ROS instead of all of the ranges defined by horizontal*vertical counts. So, I changed this in the source code to allow all of these ranges to be passed. 
To replicate before: Simply put a GPU lidar sensor into the environment and try to use the bridge in the launch file to pass the gazebo messages to ROS2. 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by mjcarroll** [comment]
Thanks for the contribution.  Two things:

1. Can you signoff your commits so that DCO will pass?

```
To add your Signed-off-by line to every commit in this branch:
Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
In your local branch, run: git rebase HEAD~2 --signoff
Force push your changes to overwrite the branch: git push --force-with-lease origin ros2
```

2. Can you add a test with some vertical elements set so that we can be sure to exercise this?

**Comment by ConnorDTaylor** [comment]
I believe I did everything requested. I added a test environment that can be run with the command: "ros2 launch ros_gz_sim_demos lidar_test.launch.py" once this is running you can then run "ros2 topic echo /lidar" and "gz topic --echo --topic /lidar" to check that the messages being passed are the same. 

If I didn't do something I thought I did please let me know! I am new to using GitHub in this capacity.

**Comment by mjcarroll** [comment]
> If I didn't do something I thought I did please let me know! I am new to using GitHub in this capacity.

Tests look good, but I think we are still having an issue with DCO.  Can you follow the instructions here and force push to get that resolved? https://github.com/gazebosim/ros_gz/pull/326/checks?check_run_id=9274824771

**Comment by ConnorDTaylor** [comment]
Will do, I did HEAD~2 from what you said, but the link says HEAD~3. So I tried that. Hopefully, it worked.

**Comment by mjcarroll** [comment]
Yep, looks good now, thanks.

File: ros_gz_sim_demos/launch/lidar_test.launch.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
Comment (ahcorde) [inline ros_gz_sim_demos/launch/lidar_test.launch.py]:
```suggestion
# Copyright 2022 Open Source Robotics Foundation, Inc.
```

File: ros_gz_sim_demos/launch/lidar_test.launch.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        PythonLaunchDescriptionSource(
+            os.path.join(pkg_ros_gz_sim, 'launch', 'gz_sim.launch.py')),
+        launch_arguments={'gz_args': '-v4 -r lidar_test.sdf' }.items(),
+    )
+
+    # RViz
+    rviz = Node(
+       package='rviz2',
+       executable='rviz2',
+       # FIXME: Generate new RViz config once this demo is usable again
```
Comment (ahcorde) [inline ros_gz_sim_demos/launch/lidar_test.launch.py]: remove traces that are not related

File: ros_gz_sim_demos/launch/lidar_test.launch.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    # Bridge
+    bridge = Node(
+        package='ros_gz_bridge',
+        executable='parameter_bridge',
+        arguments=['/lidar@sensor_msgs/msg/LaserScan[gz.msgs.LaserScan'],
+        output='screen'
+    )
+
+    # FIXME: need a SDF file (gpu_lidar.sdf) inside ros_gz_point_cloud/
```
Comment (ahcorde) [inline ros_gz_sim_demos/launch/lidar_test.launch.py]: remove

File: ros_gz_sim_demos/models/lidar_block/model.sdf [hunk 4]
```diff
@@ -0,0 +1,83 @@
+<?xml version='1.0'?>
```
Comment (ahcorde) [inline ros_gz_sim_demos/models/lidar_block/model.sdf:1]: indent the file properly

File: ros_gz_sim_demos/worlds/lidar_test.sdf [hunk 5]
```diff
@@ -0,0 +1,201 @@
+<?xml version="1.0" ?>
```
Comment (ahcorde) [inline ros_gz_sim_demos/worlds/lidar_test.sdf:1]: indent the file properly

is it the comment relevant for this demo?

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by quentingllmt** [comment]
Friendly message to show my interest. When do you plan to merge ?
Will it be backported to galactic ? 
Thanks !

**Comment by ConnorDTaylor** [comment]
> Friendly message to show my interest. When do you plan to merge ? Will it be backported to galactic ? Thanks !

I don't know why it hasn't been merged, no one has told me to fix anything and i'm not the most proficient github user so it's possible im missing something. 
I didn't plan on backporting this, I am working on Humble with Gazebo Garden. Though I I imagine fixing this problem is probably the same steps I took for this distro combination. So you should be able to just duplicate my changes. If you have any questions about it I can try to answer them, just shoot me an email.

**Review by ahcorde — COMMENTED** [review]
uncrustify is still failing https://github.com/gazebosim/ros_gz/actions/runs/3411060268/jobs/5787829085

**Comment by ahcorde** [comment]
I tested your code with the example and it's not working properly

I would expect something like this:
RViz-_002

But with your code I get:
RViz-_001

**Comment by azeey** [comment]
Looks like there are some linter issues and maybe some functionality issue here. @ConnorDTaylor have you had a chance to take a look?

**Comment by mjcarroll** [comment]
@ConnorDTaylor I thought about this a little more and realized that this approach is incorrect.  The `LaserScan` message is only intended to represent a 2d scanning lidar, like a Hokuyo or SICK sensor.  In order to get data from a 3d scanning lidar (eg velodyne/ouster), then you want to use a PointCloud based message.

The PointCloud message can additionally encode sequence/ring/intensity in the additional channels, and can maintain the structure by using a "sparse" point cloud (missing points are marked nan), or can be more space efficient by using a "dense" point cloud (no nans, but loses scan structure).

**Comment by mjcarroll** [comment]
If you have further issues, please let us know.

**Pull request closed without merging**